src/celeritas/grid/SplineXsCalculator.hh

```diff
@@ -0,0 +1,255 @@
+/*!
+ * Find and interpolate cross sections on a uniform log grid with an input
+ * spline-order.
+ *
+ * \todo Currently this is hard-coded to use "cross section grid data"
+ * which have energy coordinates uniform in log space. This should
+ * be expanded to handle multiple parameterizations of the energy grid (e.g.,
+ * arbitrary spacing needed for the Livermore sampling) and of the value
+ * interpolation (e.g. log interpolation). It might also make sense to get rid
+ * of the "prime energy" and just use log-log interpolation instead, or do a
+ * piecewise change in the interpolation instead of storing the cross section
+ * scaled by the energy.
+ *
+ * \code
+    SplineXsCalculator calc_xs(xs_grid, xs_params.reals, order);
+    real_type xs = calc_xs(particle);
+   \endcode
+ */
+class SplineXsCalculator
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using Energy = Quantity<XsGridData::EnergyUnits>;
+    using Values
+        = Collection<real_type, Ownership::const_reference, MemSpace::native>;
+    //!@}
+
+  public:
+    // Construct from state-independent data
+    inline CELER_FUNCTION SplineXsCalculator(XsGridData const& grid,
+                                             Values const& values,
+                                             size_type const order);
+
+    // Find and interpolate from the energy
+    inline CELER_FUNCTION real_type operator()(Energy energy) const;
+
+    // Get the cross section at the given index
+    inline CELER_FUNCTION real_type operator[](size_type index) const;
+
+    // Get the minimum energy
+    CELER_FUNCTION Energy energy_min() const
+    {
+        return Energy(std::exp(loge_grid_.front()));
+    }
+
+    // Get the maximum energy
+    CELER_FUNCTION Energy energy_max() const
+    {
+        return Energy(std::exp(loge_grid_.back()));
+    }
+
+  private:
+    XsGridData const& data_;
+    Values const& reals_;
+    UniformGrid loge_grid_;
+    size_type order_;
+
+    inline CELER_FUNCTION real_type get(size_type index) const;
+    CELER_FORCEINLINE_FUNCTION real_type interpolate(real_type energy,
+                                                     size_type low_idx,
+                                                     size_type high_idx) const;
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from cross section data.
+ */
+CELER_FUNCTION
+SplineXsCalculator::SplineXsCalculator(XsGridData const& grid,
+                                       Values const& values,
+                                       size_type const order)
+    : data_(grid), reals_(values), loge_grid_(data_.log_energy), order_(order)
+{
+    CELER_EXPECT(data_);
+    CELER_ASSERT(grid.value.size() == data_.log_energy.size);
+
+    // order of interpolation must be smaller than the grid size for effective
```

Comment:
For consistency we try to start comments with a capital letter.

Reply:
Fixed here and in other comments in this file